test/libponyc/flatten.cc

```diff
@@ -95,3 +95,20 @@ TEST_F(FlattenTest, TupleConstraintInUnion)
   ASSERT_EQ(2, errormsg->line);
   ASSERT_EQ(16, errormsg->pos);
 }
+
+// regression for #4016
+TEST_F(FlattenTest, TupleConstraintInUnion)
+{
+  const char* src =
+    "type Blocksize is (U8, (U8, U32))\n"
```

Comment:
I guess here we want a union type instead of a tuple of tuples?

Reply:
Yup